cmd/oras/internal/display/print.go

```diff
@@ -75,13 +76,36 @@ func PrintSuccessorStatus(ctx context.Context, desc ocispec.Descriptor, status s
 }
 
 type TagManifestStatusPrinter struct {
-	*remote.Repository
+	oras.Target
 }
 
 // PushReference overrides Repository.PushReference method to print off which tag(s) were added successfully.
 func (p *TagManifestStatusPrinter) PushReference(ctx context.Context, expected ocispec.Descriptor, content io.Reader, reference string) error {
-	if err := p.Repository.PushReference(ctx, expected, content, reference); err != nil {
-		return err
+	if repo, ok := p.Target.(registry.ReferencePusher); ok {
+		if err := repo.PushReference(ctx, expected, content, reference); err != nil {
+			return err
+		}
+	} else {
+		if err := p.Target.Tag(ctx, expected, reference); err != nil {
+			return err
+		}
 	}
```

Comment:
`TagManifestStatusPrinter` is just a wrapper and should change any business logic of the wrapper object.

Reply:
You may take a look at the implementation at https://github.com/oras-project/oras/blob/main/internal/cache/target.go#L42 to get some ideas for generating wrappers for different objects.

Reply:
The issue is: cp and tag command uses oras.TagN. It's request optimization (save 1 request) requires the target to be both `registry.ReferenceFetcher` and `registry.ReferencePusher`.